R/as-tibble.R

```diff
@@ -1,26 +1,56 @@
+#' Flatten a Crunch Cube
+#'
+#' Crunch Cubes can be expressed as a long data frame instead of a
+#' multidimensional array. In this form each dimension of the cube is a variable
+#' of the dataframe and the cube values is expressed columns for each measure.
+#' This is useful both to better understand what each entry of a cube
+#' represents, and to work with the cube result using tidyverse tools.
+#'
+#' @param x a CrunchCube
+#' @param return_real Do you want to return the tibble represenatation of the
+#'   real cube with MR selection dimensions and missing values. Mostly useful
+#'   for debugging complex cubes.
+#' @param ... futher arguments passed on to `dplyr::as_tibble()`
+#'
 #' @export
 #' @importFrom tibble as_tibble
-as_tibble.CrunchCube <- function (x, ...) {
+as_tibble.CrunchCube <- function (x, return_real = FALSE, ...) {
     ## TODO: Consider using `dplyr::tbl_cube` class
     ## TODO: NA behavior? Add way to get missings not missing. (This uses
     ## as.array to remove extraneous dims, and doing so applies the '@useNA'
     ## from the cube)
     ## TODO: deal with naming for array (and NUMR) variable dims (var.categories?)
-
+    
     ## 1) get the measures
     ## TODO: better access methods for secondary measures in Cube
     ## TODO: if weighted query, include ".unweighted_counts"
     measure_names <- setdiff(names(x@arrays), ".unweighted_counts")
-    vals <- sapply(measure_names, function (m) {
-        as.vector(crunch:::cubeToArray(x, m))
-    }, simplify=FALSE)
-
-    ## 2) If there are dimnames, expand.grid and c(bind) them
-    dnames <- dimnames(as.array(x))
-    ## NULL if scalar value, which means no "group_by"
-    if (!is.null(dnames)) {
-        dims <- do.call(expand.grid, dnames)
-        vals <- c(dims, vals)
+    if (return_real) {
+        is_selected <- crunch:::is.selectedDimension(x@dims)
+        dnames <- dimnames(x@arrays$count)
+        out <- do.call(expand.grid, dnames)
+        if (any(is_selected)) {
+            names(out)[is_selected] <- paste0(names(out)[is_selected], "_selections")
+        }
```

Comment:
This is great for MR, but what happens with categorical arrays? Those will have the same name as well even though one dimension is items and the other is categories.